oauth2-useragent-core/src/main/java/com/microsoft/alm/oauth2/useragent/UserAgentImpl.java

```diff
@@ -157,70 +205,69 @@ void addClassPathToCommand(final List<String> classPath, final List<String> comm
         command.add("-classpath");
         //noinspection ToArrayCallWithZeroLengthArrayArgument
         final String[] classPathComponents = classPath.toArray(EMPTY_STRING_ARRAY);
-        final String classPathString = join(pathSeparator, classPathComponents);
+        final String classPathString = StringHelper.join(pathSeparator, classPathComponents);
         command.add(classPathString);
     }
 
-    static String join(final String separator, final String[] value)
-    {
-        if (value == null)
-            throw new IllegalArgumentException("value is null");
-
-        // "If separator is null, an empty string (String.Empty) is used instead."
-        final String sep = separator == null ? "" : separator;
+    static void throwUnsupported(final Map<Provider, List<String>> unmetRequirements) {
+        final StringBuilder sb = new StringBuilder("I don't support your platform yet.");
+        describeUnmetRequirements(unmetRequirements, sb);
+        sb.append(NEW_LINE);
+        sb.append("Please send details about your operating system version, Java version, 32- vs. 64-bit, etc.");
+        sb.append(NEW_LINE);
+        sb.append("The following System Properties and Environment Variables would be very useful.");
+        sb.append(NEW_LINE);
 
-        final StringBuilder result = new StringBuilder();
+        final Properties properties = System.getProperties();
+        appendProperties(properties, sb);
+        sb.append(NEW_LINE);
 
-        if (value.length > 0) {
-            result.append(value[0] == null ? "" : value[0]);
-            for (int i = 1; i < value.length; i++) {
-                result.append(sep);
-                result.append(value[i] == null ? "" : value[i]);
-            }
-        }
+        final Map<String, String> variables = System.getenv();
+        appendVariables(variables, sb);
 
-        return result.toString();
+        throw new IllegalStateException(sb.toString());
     }
 
-    static Provider determineProvider(final String userAgentProvider) {
-        return determineProvider(userAgentProvider, Provider.PROVIDERS);
-    }
+    static Provider scanProviders(final String userAgentProvider, final List<Provider> providers, final Map<Provider, List<String>> destinationUnmetRequirements) {
 
-    static Provider determineProvider(final String userAgentProvider, final List<Provider> providers) {
+        Provider result = null;
 
         if (userAgentProvider != null) {
             for (final Provider provider : providers) {
                 if (provider.getClassName().equals(userAgentProvider)) {
-                    return provider;
+                    result = provider;

```

Comment:
> Otherwise, once you're deemed compatible (or incompatible), it's unlikely to stop (or start) working, so there's no need to keep checking.

It's understood that there is no need to keep checking, the comment is more about code structure -- remove states IMHO will make the code easier to read.  

"Expensive" is relative, the process we are about to invoke is a user interaction, which could runs in minutes.  As long as the check is tolerable (sub second?), it may not matter that much.  

Another option is instead of "true/false", how about determineProvider returns the compatible provider, and then consumer can pass this `Provider` to `requestAuthorizationCode`, or create a new function named `requestAuthorizationCodeWithProvider` and just use it.  This way you might be able to avoid storing state?